File: tk-product-research/scripts/calculate_product_scores.js
#!/usr/bin/env node
const fs = require("fs");

function parseCsv(text) {
  const lines = text.trim().split(/\r?\n/);
  const headers = parseCsvLine(lines.shift()).map((h) => h.trim());
  return lines.filter(Boolean).map((line) => {
    const values = parseCsvLine(line).map((v) => v.trim());
    return Object.fromEntries(headers.map((h, i) => [h, values[i] ?? ""]));
  });
}

function parseCsvLine(line) {
  const values = [];
  let current = "";
  let quoted = false;
  for (let i = 0; i < line.length; i += 1) {
    const char = line[i];
    const next = line[i + 1];
    if (char === '"' && quoted && next === '"') {
      current += '"';
      i += 1;
    } else if (char === '"') {
      quoted = !quoted;
    } else if (char === "," && !quoted) {
      values.push(current);
      current = "";
    } else {
      current += char;
    }
  }
  values.push(current);
  return values;
}

function toCsv(rows) {
  const headers = [...new Set(rows.flatMap((row) => Object.keys(row)))];
  const escape = (value) => {
    const text = String(value ?? "");
    return /[",\n]/.test(text) ? `"${text.replace(/"/g, '""')}"` : text;
  };
  return [headers.join(","), ...rows.map((row) => headers.map((h) => escape(row[h])).join(","))].join("\n");
}

function number(row, key, fallback = 0) {
  const raw = row[key];
  if (raw === undefined || raw === "") return fallback;
  const parsed = Number(String(raw).replace("%", ""));
  return Number.isFinite(parsed) ? parsed : fallback;
}

function rate(value) {
  return value > 1 ? value / 100 : value;
}

function marginScore(value) {
  if (value >= 0.55) return 90;
  if (value >= 0.40) return 70;
  if (value >= 0.25) return 50;
  return 25;
}

function priceFit(price) {
  if (price <= 0) return 50;
  if (price <= 20) return 90;
  if (price <= 40) return 75;
  if (price <= 80) return 55;
  return 35;
}

function score(row) {
  const sellingPrice = number(row, "selling_price");
  const productCost = number(row, "product_cost");
  const shippingCost = number(row, "shipping_cost");
  const commissionRate = rate(number(row, "commission_rate", 0.12));
  const grossMargin = sellingPrice - productCost - shippingCost - sellingPrice * commissionRate;
  const grossMarginRate = sellingPrice ? grossMargin / sellingPrice : 0;
  const breakevenRoas = grossMargin > 0 ? sellingPrice / grossMargin : 0;
  const competitionScore = number(row, "competition_score", 50);
  const viralProbability = Math.max(0, Math.min(100,
    number(row, "trend_score", 50) * 0.30 +
    number(row, "content_score", 50) * 0.20 +
    priceFit(sellingPrice) * 0.15 +
    marginScore(grossMarginRate) * 0.15 +
    number(row, "supplier_score", 50) * 0.10 -
    competitionScore * 0.10
  ));
  const recommendation = viralProbability >= 70 && grossMarginRate >= 0.40
    ? "test"
    : viralProbability >= 55 || grossMarginRate >= 0.35 ? "watch" : "avoid";
  return {
    ...row,
    gross_margin: grossMargin.toFixed(2),
    gross_margin_rate: grossMarginRate.toFixed(4),
    breakeven_roas: breakevenRoas.toFixed(2),
    viral_probability: viralProbability.toFixed(1),
    competition_strength: competitionScore.toFixed(1),
    recommendation,
  };
}

const [input, output] = process.argv.slice(2);
if (!input || !output) {
  console.error("Usage: node calculate_product_scores.js input.csv output.csv");
  process.exit(1);
}

fs.writeFileSync(output, toCsv(parseCsv(fs.readFileSync(input, "utf8")).map(score)), "utf8");

File: versions/tk-skills-v1-20260624/tk-product-research/scripts/calculate_product_scores.py
#!/usr/bin/env python3
import argparse
import csv
from pathlib import Path


def as_float(row, key, default=0.0):
    value = row.get(key, default)
    if value in (None, ""):
        return default
    try:
        return float(str(value).strip().rstrip("%"))
    except ValueError:
        return default


def clamp(value, low=0.0, high=100.0):
    return max(low, min(high, value))


def margin_score(rate):
    if rate >= 0.55:
        return 90
    if rate >= 0.40:
        return 70
    if rate >= 0.25:
        return 50
    return 25


def price_fit_score(price):
    if price <= 0:
        return 50
    if price <= 20:
        return 90
    if price <= 40:
        return 75
    if price <= 80:
        return 55
    return 35


def normalize_rate(value):
    return value / 100 if value > 1 else value


def score_row(row):
    selling_price = as_float(row, "selling_price")
    product_cost = as_float(row, "product_cost")
    shipping_cost = as_float(row, "shipping_cost")
    commission_rate = normalize_rate(as_float(row, "commission_rate", 0.12))
    platform_fee = selling_price * commission_rate
    gross_margin = selling_price - product_cost - shipping_cost - platform_fee
    gross_margin_rate = gross_margin / selling_price if selling_price else 0
    breakeven_roas = selling_price / gross_margin if gross_margin > 0 else 0

    trend_score = as_float(row, "trend_score", 50)
    content_score = as_float(row, "content_score", 50)
    competition_score = as_float(row, "competition_score", 50)
    supplier_score = as_float(row, "supplier_score", 50)

    viral_probability = (
        trend_score * 0.30
        + content_score * 0.20
        + price_fit_score(selling_price) * 0.15
        + margin_score(gross_margin_rate) * 0.15
        + supplier_score * 0.10
        - competition_score * 0.10
    )
    viral_probability = clamp(viral_probability)

    if viral_probability >= 70 and gross_margin_rate >= 0.40:
        recommendation = "test"
    elif viral_probability >= 55 or gross_margin_rate >= 0.35:
        recommendation = "watch"
    else:
        recommendation = "avoid"

    return {
        **row,
        "gross_margin": round(gross_margin, 2),
        "gross_margin_rate": round(gross_margin_rate, 4),
        "breakeven_roas": round(breakeven_roas, 2),
        "viral_probability": round(viral_probability, 1),
        "competition_strength": round(competition_score, 1),
        "recommendation": recommendation,
    }


def main():
    parser = argparse.ArgumentParser(description="Score TikTok Shop product candidates.")
    parser.add_argument("input_csv", type=Path)
    parser.add_argument("output_csv", type=Path)
    args = parser.parse_args()

    with args.input_csv.open("r", newline="", encoding="utf-8-sig") as source:
        rows = list(csv.DictReader(source))

    scored = [score_row(row) for row in rows]
    fieldnames = list(dict.fromkeys([key for row in scored for key in row.keys()]))

    with args.output_csv.open("w", newline="", encoding="utf-8") as target:
        writer = csv.DictWriter(target, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(scored)


if __name__ == "__main__":
    main()
